sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,419 @@
+                            if (!string.IsNullOrEmpty(secretEnvVar))
+                            {
+                                s_secret = secretEnvVar;
+                                s_msiType = MsiType.AppService;
+                            }
+                            // if ONLY the env var MSI_ENDPOINT is set the MsiType is CloudShell
+                            else
+                            {
+                                s_msiType = MsiType.AppService;
+                            }
+                        }
+                        // if MSI_ENDPOINT is NOT set AND the IMDS endpoint is available the MsiType is Imds
+                        else if (ImdsAvailable(cancellationToken))
+                        {
+                            s_endpoint = ImdsEndpoint;
+                            s_msiType = MsiType.Imds;
+                        }
+                        // if MSI_ENDPOINT is NOT set and IMDS enpoint is not available ManagedIdentity is not available
+                        else
+                        {
+                            s_msiType = MsiType.Unavailable;
+                        }
+                    }
+                }
+                // release the init lock
+                finally
+                {
+                    s_initLock.Release();
+                }
+            }
+
+            return s_msiType;
+        }
+
+        private bool ImdsAvailable(CancellationToken cancellationToken)
+        {            
+            // send a request without the Metadata header.  This will result in a failed request,
+            // but we're just interested in if we get a response before the timeout of 500ms
+            // if we don't get a response we assume the imds endpoint is not available
+            using (Request request = _pipeline.CreateRequest())
+            {
+                request.Method = HttpPipelineMethod.Get;
+
+                request.UriBuilder.Uri = ImdsEndpoint;
+
+                request.UriBuilder.AppendQuery("api-version", ImdsApiVersion);
+
+                var imdsTimeout = new CancellationTokenSource(500).Token;
+
+                try
+                {
+                    var response = _pipeline.SendRequest(request, CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, imdsTimeout).Token);
+                
+                    return true;
+                }
+                // we only want to handle the case when the imdsTimeout resulted in the request being cancelled.
+                // this indicates that the request timed out and that imds is not available.  Otherwise the operation
+                // was user cancelled.  In this case we don't wan't to handle the exception so s_identityAvailable will
+                // remain unset, as we still haven't determined if the imds endpoint is available.
+                catch (OperationCanceledException ex) when (ex.CancellationToken == imdsTimeout)
```

Comment:
Are you sure ex.CancellationToken will be ever imdsTimeout? I thought it would be the combined token.

Reply:
Good point.  Rather than comparing ex.CancellationToken to imdsTimeout I could check imdsTimout.IsCancellationReqeusted in the filter.  This does have a race in it when cancellationToken is responsible for cancelling the requests and then imdsTimout expires after but before the filter is executed, but perhaps this is acceptable.  Thoughts?

Reply:
can you check `cancellationToken.IsCancellationRequested`?